diff '--color=auto' -Naur a/setup.py b/setup.py
--- a/setup.py	2024-05-11 21:26:47.522949960 +0200
+++ b/setup.py	2024-05-11 21:27:09.206583286 +0200
@@ -93,7 +93,7 @@
                 language="c++",
                 include_dirs=[os.path.join(re2_prefix, "include")] if re2_prefix else [],
                 libraries=["re2"],
-                extra_compile_args=['-std=c++11'],
+                extra_compile_args=['-std=c++17'],
                 library_dirs=[os.path.join(re2_prefix, "lib")] if re2_prefix else [],
                 runtime_library_dirs=[os.path.join(re2_prefix, "lib")] if re2_prefix else [],
             )
